llvm/lib/IR/RepoTicket.cpp

```diff
@@ -135,26 +165,21 @@ static bool updateDigestUseDependenciesAndContributions(
     return true;
   }
 
-  bool Changed = false;
   GOHash.update('T');
   auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  Changed = Changed || State.Changed;
-  // Update the GO's hash using its contributions and dependencies.
-  auto &GOContributions = State.Contributions;
-  auto &GODependencies = State.Dependencies;
-  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
-           make_range(GOContributions.begin(), GOContributions.end()),
-           make_range(GODependencies.begin(), GODependencies.end()))) {
-    const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
-    // if function will not be inlined and not be discarded if it is not used,
-    // skip it.
-    if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
-        !Fn->isDiscardableIfUnused())
-      continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
-  }
+  bool Changed = State.Changed || !State.Contributions.empty();
+  // The hashes of the GO's 'contributions' and 'dependencies' are separately
+  // added to the GO's hash since their respective arrows point are in different
+  // direction. In addition, 'contributions' cannot form loops since 1) only
+  // global variables have 'contributions' and 2) all 'contributions' are
+  // functions. Therefore, the 'contributions' can be simplified not to record
+  // 'visited' table, which avoids having two 'visited' maps. Firstly, add the
+  // GO's contributions to its hash.
+  updateDigestUseContributions(GO, GOHash, State.Contributions, GOIMap);
```

Comment:
Related to Paul's comments above, I feel like the behaviour of Changed is fairly obscure here.

Maybe this could be:
```
bool Changed = State.Changed;
Changed = Changed || updateDigestUseContributions(GO, GOHash, State.Contributions, GOIMap);
Changed = Changed || updateDigestUseDependencies(GOHash, State.Dependencies, Visited, GOIMap,
                        AccumulateGODigest);
return Changed
```

Moving the State.Contributions.empty() into updateDigestUseContributions, and changing updateDigestUseDependencies to return Changed.

Reply:
> ```
> Changed = Changed || updateDigestUseContributions(GO, GOHash, State.Contributions, GOIMap);
> Changed = Changed || updateDigestUseDependencies(GOHash, State.Dependencies, Visited, GOIMap, AccumulateGODigest);
> ```
> Moving the State.Contributions.empty() into updateDigestUseContributions, and changing updateDigestUseDependencies to return Changed.

Thanks, I have modified the code following your advice. However, I changed 
```
Changed = updateDigestUseContributions(GO, GOHash, State.Contributions, GOIMap) || Changed;
Changed = updateDigestUseDependencies(GOHash, State.Dependencies, Visited, GOIMap, AccumulateGODigest) || Changed ;
```
If `Changed` is put in the beginning and its value is `true`, the functions of `updateDigestUseContributions` and `updateDigestUseDependencies` will not be called.
Therefore, I reorder the two items.

Reply:
> If Changed is put in the beginning and its value is true, the functions of updateDigestUseContributions and updateDigestUseDependencies will not be called.

Good point!